### PR TITLE
new feature start, saving rdf skeleton on disk for reloading

### PR DESCRIPTION
**Pull request by Shamanou**
Added option for the user to save the rdf skeleton with file format metadata to the harddisk.

The files are stored in the same location as the project files used by openrefine.
this is by default: 
~/.local/share/openrefine/

### CODE REVIEW

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
+package org.dtls.fairifier;
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: our domain is 'nl.dtls'

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class FileSystemRdfSkeletonImp implements RdfSkeletonService{
+    private static final Path SAVELOCATION = Paths.get(System.getProperty("user.home") + "/.local/share/openrefine/");
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: Is there an existing storage location available through openrefine? The location where the projects themselves are persisted maybe?

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class FileSystemRdfSkeletonImp implements RdfSkeletonService{
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: Missing a whitespace before the `{`.

Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: I'd prefer the `Impl` suffix convention, but that's up for discussion.

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    @Override
+    public List<String> listModels(String fileType) {
+        File folder = SAVELOCATION.toFile();
+        ArrayList<String> out = new ArrayList<String>();
+        File[] listOfFiles = folder.listFiles();
+        for (int i = 0; i < listOfFiles.length; i++) {
+            if (listOfFiles[i].isFile()) {
+                String name = listOfFiles[i].getName().split(".")[0];
+                try {
+                    ObjectMapper mapper = new ObjectMapper();
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: An `ObjectMapper` instance is threadsafe after configuration, so we can use a single instance for this class.

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+public class FileSystemRdfSkeletonImp implements RdfSkeletonService{
+    private static final Path SAVELOCATION = Paths.get(System.getProperty("user.home") + "/.local/share/openrefine/");
+    private final static Logger logger = LoggerFactory.getLogger("FileSystemRdfSkeletonImp");
+
+    @Override
+    public List<String> listModels(String fileType) {
+        File folder = SAVELOCATION.toFile();
+        ArrayList<String> out = new ArrayList<String>();
+        File[] listOfFiles = folder.listFiles();
+        for (int i = 0; i < listOfFiles.length; i++) {
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: If we have java 8 available, we can use [`Files.list`](http://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#list-java.nio.file.Path-) to simplify lines 22 through 27.

If we have java 7 available we can use [`Files.walkFileTree`](http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#walkFileTree(java.nio.file.Path,%20java.util.Set,%20int,%20java.nio.file.FileVisitor)) to do roughly the same with a bit more effort.

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            if (listOfFiles[i].isFile()) {
+                String name = listOfFiles[i].getName().split(".")[0];
+                try {
+                    ObjectMapper mapper = new ObjectMapper();
+                    SkeletonMetadata metadata = mapper.readValue(new File(listOfFiles[i].getParent() + File.separator + name + "metadata.skeleton.json"), SkeletonMetadata.class);
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: If we manage to simplify/modernise the file listing, we can probably simplify this as well. At least the creation of the file or path object, but maybe the conversion step as well. Up for discussion once we figure out the file listing.

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                    if (metadata.getFileType().equals(fileType)) {
+                        out.add(new String(Files.readAllBytes(Paths.get(listOfFiles[i].getPath()))));
+                    }
+                } catch (IOException e) {
+                    logger.error(e.getMessage());
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: Is this communicated back to the front-end? How will the user know if something goes wrong while saving?

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                }
+            }
+        }
+        return out;
+    }
+
+    @Override
+    public List<String> listModels() {
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: Same argument for the implementation of this method: i think we can simplify this method by using more modern approaches.

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                }
+            }
+        }
+        return out;
+    }
+    
+    @Override
+    public void saveModel(String json, String fileType, String projectId) {
+        try {
+            BufferedWriter out = new BufferedWriter(new FileWriter(SAVELOCATION.toString() + File.separator + projectId + ".skeleton.json"));
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: I think we can simplify the writer handling. If we have java 8 available, we can use [`Files.newBufferedWriter`](http://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#newBufferedWriter-java.nio.file.Path-java.nio.charset.Charset-java.nio.file.OpenOption...-) in a try-with-resources. And we can also reuse a  base `Path` for writing both files.

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            SkeletonMetadata metadata = new SkeletonMetadata();
+            metadata.setFileType(fileType);
+            mapper.writeValue(new File(SAVELOCATION.toString() + File.separator  + projectId + ".metadata.skeleton.json"), metadata);
+        } catch (IOException e) {
+            logger.error(e.getMessage());
+        }
+    }
+
+    @Override
+    public String loadModel(String projectId) {
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: Same argument for the implementation of this method: i think we can simplify this method by using more modern approaches.

File: extensions/fairifier-rdf-extension/dtls/fairifier/RdfSkeletonService.java [hunk 11]
```diff
@@ -0,0 +1,10 @@
+package org.dtls.fairifier;
+
+import java.util.List;
+
+interface RdfSkeletonService {
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/RdfSkeletonService.java]: Needs more javadoc.

File: extensions/fairifier-rdf-extension/dtls/fairifier/SaveRdfSkeletonCommand.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+import java.io.IOException;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.refine.commands.Command;
+
+public class SaveRdfSkeletonCommand extends Command{
+    private static final RdfSkeletonService rdfSkeletonService = new FileSystemRdfSkeletonImp();
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/dtls/fairifier/SaveRdfSkeletonCommand.java]: Doesn't need to be a static member?

File: extensions/fairifier-rdf-extension/scripts/rdf-schema-alignment.js [hunk 13]
```diff
@@ -113,6 +113,10 @@ RdfSchemaAlignmentDialog.prototype._constructBody = function(body) {
                     }
                 }
             );
+        var fileType = window.prompt("filetype","");
+        $.post("command/rdf-extension/save-rdf-skeleton",{model: JSON.stringify(self.getJSON()), projectId: theProject.id, fileType: fileType},function(data){
```
Comment (kburger) [inline extensions/fairifier-rdf-extension/scripts/rdf-schema-alignment.js]: Maybe we can make this a bit more readable with newlines and indentation, the oneliner could be hard to read.

**Review by kburger — CHANGES_REQUESTED** [review]
Room for improvement, especially on using more modern ways of handling files.

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java (hunk 1 above)
Comment (Shamanou) [reply extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: The fairifier now uses org, but i could change that

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java (hunk 2 above)
Comment (Shamanou) [reply extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: That would be this location, but maybe i could also pull it from the configs

File: extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java (hunk 7 above)
Comment (Shamanou) [reply extensions/fairifier-rdf-extension/dtls/fairifier/FileSystemRdfSkeletonImp.java]: good point ill change it

**Review by kburger — APPROVED** [review]

**Pull request merged**